From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Thu, 18 Dec 2025 23:16:03 +0800
Subject: [PATCH 20/47] sep: Abstract SEP messages from the underlying mailbox

Abstract SEP message from the underlying mailbox such that support for
SEP behind different types of mailbox can be added.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 src/sep.c | 79 ++++++++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 67 insertions(+), 12 deletions(-)

diff --git a/src/sep.c b/src/sep.c
index 7a40fef2..8be0fc1c 100644
--- a/src/sep.c
+++ b/src/sep.c
@@ -2,7 +2,9 @@
 
 #include <string.h>
 
+#include "adt.h"
 #include "asc.h"
+#include "malloc.h"
 #include "sep.h"
 #include "types.h"
 #include "utils.h"
@@ -18,21 +20,74 @@
 
 #define SEP_TIMEOUT 1000
 
-static asc_dev_t *sep_asc = NULL;
+enum sep_mbox_type {
+    SEP_MBOX_TYPE_ASC,
+};
+
+static struct sep_dev {
+    enum sep_mbox_type type;
+    union {
+        asc_dev_t *asc;
+    };
+} *sep_dev;
 
 int sep_init(void)
 {
-    if (!sep_asc)
-        sep_asc = asc_init("/arm-io/sep");
-    if (!sep_asc)
+    const char *path = "/arm-io/sep";
+    int sep_path[8];
+
+    int node = adt_path_offset_trace(adt, path, sep_path);
+    if (node < 0) {
+        printf("sep: Error getting sep node %s\n", path);
+        return -1;
+    }
+
+    u64 base;
+    if (adt_get_reg(adt, sep_path, "reg", 0, &base, NULL) < 0) {
+        printf("sep: Error getting akf %s base address.\n", path);
         return -1;
+    }
+
+    sep_dev = calloc(0, sizeof(*sep_dev));
+    sep_dev->type = SEP_MBOX_TYPE_ASC;
+
+    switch (sep_dev->type) {
+        case SEP_MBOX_TYPE_ASC:
+            sep_dev->asc = asc_init(path);
+            break;
+    }
+
     return 0;
 }
 
+bool sep_send(u64 msg)
+{
+    switch (sep_dev->type) {
+        case SEP_MBOX_TYPE_ASC: {
+            const struct asc_message asc_msg = {.msg0 = msg};
+            return asc_send(sep_dev->asc, &asc_msg);
+        }
+    }
+    __builtin_unreachable();
+}
+
+bool sep_recv(u64 *reply)
+{
+    switch (sep_dev->type) {
+        case SEP_MBOX_TYPE_ASC: {
+            struct asc_message asc_reply;
+            int retval = asc_recv_timeout(sep_dev->asc, &asc_reply, SEP_TIMEOUT);
+            *reply = asc_reply.msg0;
+            return retval;
+        }
+    }
+    __builtin_unreachable();
+}
+
 size_t sep_get_random(void *buffer, size_t len)
 {
-    const struct asc_message msg_getrand = {.msg0 = FIELD_PREP(SEP_MSG_EP, SEP_EP_ROM) |
-                                                    FIELD_PREP(SEP_MSG_CMD, SEP_MSG_GETRAND)};
+    const u64 msg_getrand =
+        FIELD_PREP(SEP_MSG_EP, SEP_EP_ROM) | FIELD_PREP(SEP_MSG_CMD, SEP_MSG_GETRAND);
     int ret;
     size_t done = 0;
 
@@ -41,20 +96,20 @@ size_t sep_get_random(void *buffer, size_t len)
         return 0;
 
     while (len) {
-        struct asc_message reply;
+        u64 reply;
         u32 rng;
         size_t copy;
 
-        if (!asc_send(sep_asc, &msg_getrand))
+        if (!sep_send(msg_getrand))
             return done;
-        if (!asc_recv_timeout(sep_asc, &reply, SEP_TIMEOUT))
+        if (!sep_recv(&reply))
             return done;
-        if (FIELD_GET(SEP_MSG_CMD, reply.msg0) != SEP_REPLY_GETRAND) {
-            printf("SEP: unexpected getrand reply: %016lx\n", reply.msg0);
+        if (FIELD_GET(SEP_MSG_CMD, reply) != SEP_REPLY_GETRAND) {
+            printf("SEP: unexpected getrand reply: %016lx\n", reply);
             return done;
         }
 
-        rng = FIELD_GET(SEP_MSG_DATA, reply.msg0);
+        rng = FIELD_GET(SEP_MSG_DATA, reply);
         copy = sizeof(rng);
         if (copy > len)
             copy = len;
